compiler/src/Language/Bond/Codegen/Java/Class_java.hs

```diff
@@ -473,6 +473,70 @@ object_hashCode fields structBase = [lt|@Override
         hashCode _ f          = [lt|#{f} == null ? 0 : #{f}.hashCode()|]
 
 
+-- We implement Externalizable, rather than Serializable, so that
+-- ser/deserialization will result in a single call on the most derived class,
+-- rather than one call for each type in the inheritance chain. By reading into
+-- a byte[] instead of trying to deserialize from the ObjectInput{,Stream} we're
+-- passed, we can start from a ByteArrayInputStream, which we already know how
+-- to clone for Bonded fields.
+--
+-- We write the length of the serialized data because we can't assume the
+-- ObjectInput{,Stream} actually ends after the object we care about.
+--
+-- serialVersionUID is always 0 so that Java will always delegate compatibility
+-- checking of serialized data against current deserialization code to us.
+javaNativeSerializationGlue :: String -> Text
+javaNativeSerializationGlue declName = [lt|
+    // Java native serialization
+    private static final long serialVersionUID = 0L;
+    private #{declName} __deserializedInstance;
+
+    @Override
+    public void writeExternal(java.io.ObjectOutput out) throws java.io.IOException {
+        final java.io.ByteArrayOutputStream outStream = new java.io.ByteArrayOutputStream();
+        final org.bondlib.ProtocolWriter writer = new org.bondlib.CompactBinaryWriter(outStream, 2);
+        org.bondlib.Marshal.marshal(this, writer);
+
+        final byte[] marshalled = outStream.toByteArray();
+        out.writeInt(marshalled.length);
+        out.write(marshalled);
+    }
+
+    @Override
+    public void readExternal(java.io.ObjectInput in) throws java.io.IOException, java.lang.ClassNotFoundException {
+        final int marshalledLength = in.readInt();
+        final byte[] marshalled = new byte[marshalledLength];
+        final int bytesRead = in.read(marshalled);
```

Comment:
From the API documentation for [`ObjectInput.read`](http://docs.oracle.com/javase/7/docs/api/java/io/ObjectInput.html#read(byte[])), it looks like `read` needs to be called in a loop until the requested number of bytes have been read or EOF is signaled.

Reply:
I don't get that implication from that doc, but I dug into this mess and found that [`DataInput.readFully()`](http://docs.oracle.com/javase/6/docs/api/java/io/DataInput.html#readFully(byte[])) guarantees the behavior you want. `ObjectInput extends DataInput`, though the doc that suggests it is only in the method of `ObjectInputStream` with the [exact same signature](http://docs.oracle.com/javase/6/docs/api/java/io/ObjectInputStream.html#read(byte[],%20int,%20int)).

☕️🤷‍♂️